### PR TITLE
Revert "Ensure there is one running static pod with the same full name"

### PR DESCRIPTION
**Pull request by rphillips**
Reverts kubernetes/kubernetes#104743 https://github.com/kubernetes/kubernetes/pull/107695

Issue: https://github.com/kubernetes/kubernetes/issues/107733

```release-note
Reverts handling of static pod fullnames that may sometimes cause static pods to fail to restart.
```

### CODE REVIEW

**Comment by ehashman** [comment]
I think the revert makes sense.

I noticed when I was trying to add coverage for https://github.com/kubernetes/kubernetes/pull/107695 that we added `newStaticPod` here in the test coverage but not `newStaticPodWithPhase` to match the existing mocks: 


We lack coverage for this: https://github.com/kubernetes/kubernetes/blob/5426da8f69c1d5fa99814526c1878aeb99b2456e/pkg/kubelet/pod_workers.go#L572-L576

So @rphillips tried to add in the missing pod fullname, but I'm not 100% sure we have a reproducer that can catch that.

I started adding a case:

```diff
diff --git a/pkg/kubelet/pod_workers_test.go b/pkg/kubelet/pod_workers_test.go
index 4028c06c292..fc5f975acfd 100644
--- a/pkg/kubelet/pod_workers_test.go
+++ b/pkg/kubelet/pod_workers_test.go
@@ -164,6 +164,21 @@ func newStaticPod(uid, name string) *v1.Pod {
        }
 }
 
+func newStaticPodWithPhase(uid, name string, phase v1.PodPhase) *v1.Pod {
+       return &v1.Pod{
+               ObjectMeta: metav1.ObjectMeta{
+                       UID:  types.UID(uid),
+                       Name: name,
+                       Annotations: map[string]string{
+                               kubetypes.ConfigSourceAnnotationKey: kubetypes.FileSource,
+                       },
+               },
+               Status: v1.PodStatus{
+                       Phase: phase,
+               },
+       }
+}
+
 // syncPodRecord is a record of a sync pod call
 type syncPodRecord struct {
        name       string
@@ -856,6 +871,24 @@ func Test_allowPodStart(t *testing.T) {
                        },
                        allowed: true,
                },
+               {
+                       desc: "static pod if the static pod is terminated and rebooting",
+                       pod:  newStaticPodWithPhase("uid-0", "foo", v1.PodSucceeded),
+                       podSyncStatuses: map[types.UID]*podSyncStatus{
+                               "uid-0": {
+                                       fullname: "foo_",
+                               },
+                       },
+                       // waitingToStartStaticPodsByFullname: map[string][]types.UID{
+                       //      "foo_": {
+                       //              types.UID("uid-0"),
+                       //      },
+                       // },
+                       startedStaticPodsByFullname: map[string]types.UID{
+                               "foo_": types.UID("uid-0"),
+                       },
+                       allowed: true,
+               },
        }
 
        for _, tc := range testCases {
```

But then I realized that we don't have anything mocking the worker [podCache](https://github.com/kubernetes/kubernetes/blob/5426da8f69c1d5fa99814526c1878aeb99b2456e/pkg/kubelet/pod_workers.go#L575), which will be necessary to give proper coverage, and so we probably have a whole matrix of missing test cases. Hence, I think it might be easiest to revert for now.

/lgtm
/cc @gjkim42 
/triage accepted

/release-note-edit
```release-note
Reverts handling of static pod fullnames that may sometimes cause static pods to fail to restart.
```

**Comment by liggitt** [comment]
The original commit was also backported to 1.22 (https://github.com/kubernetes/kubernetes/pull/106394). Will this revert also be backported? Does this reintroduce any regressions to the 1.22 and 1.23 streams?

**Review by gjkim42 — COMMENTED** [review]
I think we need to clarify the problematic scenario for this. And then we can ensure it will be addressed by reverting this.

I guess [this](https://github.com/kubernetes/kubernetes/pull/104847) may be relevant to the problem.

**Comment by rphillips** [comment]
We are going to revert the entire PR: https://github.com/kubernetes/kubernetes/pull/104743

**Comment by ehashman** [comment]
> The original commit was also backported to 1.22 (#106394). Will this revert also be backported? Does this reintroduce any regressions to the 1.22 and 1.23 streams?

If we revert, I think the reverts should be backported.

The patch was intended to fix https://github.com/kubernetes/kubernetes/issues/97722#issuecomment-888717438 which sounds like the current failure mode we're seeing in OpenShift, so that fix did not work in all cases. We discovered that we are still seeing some affected workloads in https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/26796/pull-ci-openshift-origin-master-e2e-aws-serial/1487126569716551680

Taking https://github.com/kubernetes/kubernetes/pull/107695 was low risk and did help, but per my concerns in https://github.com/kubernetes/kubernetes/pull/107734#issuecomment-1020489430 I think that we missed an entire matrix of test coverage because code that was not changed was affected by this change and missed in review.

Given all of this, our options are:
- revert
- try to figure out a fix forward

The problem is, I'm not sure what a quick forward fix would be :)

**Comment by gjkim42** [comment]
First I can't dig this issue as there is no clear reproducer.

> https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/26796/pull-ci-openshift-origin-master-e2e-aws-serial/1487126569716551680

I am not sure what this CI job was doing, but I guess the CI job was trying to restart an etcd pod which is a static pod. Am I right?
Does the job always fail? or sometimes?


> Given all of this, our options are:
> 
> * revert
> * try to figure out a fix forward
> 
> The problem is, I'm not sure what a quick forward fix would be :)

**_I think at least we have to confirm that this PR(reverting) can address the problem._**

**Comment by gjkim42** [comment]
Found a reproducer here. 
https://github.com/kubernetes/kubernetes/issues/107733#issuecomment-1024895999
e2e test: https://github.com/kubernetes/kubernetes/pull/107855

And made a fix for it.
https://github.com/kubernetes/kubernetes/pull/107854

Maybe the fix could be an alternative.

**Comment by rphillips** [comment]
/hold cancel

+1 with @ehashman ... We need to revisit the original PR and get more testing around it.

**Review by mrunalp — APPROVED** [review]

**Comment by smarterclayton** [comment]
/hold

So one serious concern here is that the fix was a demonstrable hard break in static pods.  If we've regressed elsewhere, reverting just moves the problem around (doesn't result in fewer problems).  I'd prefer to spend a small amount of time and find a fix, while at the same time improving the test coverage based on our new failure mode.

**Comment by ehashman** [comment]
> So one serious concern here is that the fix was addressing a demonstrable hard break in static pods. If we've regressed elsewhere, reverting just moves the problem around (doesn't result in fewer problems). I'd prefer to spend a small amount of time and find a fix, while at the same time improving the test coverage based on our new failure mode.

I believe the demonstrable break was your PR https://github.com/kubernetes/kubernetes/pull/104847 which fixed https://github.com/kubernetes/kubernetes/issues/104648. I think there may be some mixed communication around bugs and patches which caused this patch to land; it is not clear to me what this is actually fixing, and it's definitely causing a regression.

/test pull-kubernetes-e2e-gce-ubuntu-containerd

**Comment by rphillips** [comment]
+1 the original PR clearly regresses the static pods, and the reasoning behind the issue may have been fixed in the followup pod lifecycle PRs we fixed last year.

**Comment by smarterclayton** [comment]
https://github.com/kubernetes/kubernetes/pull/104743#issuecomment-915332807 and https://github.com/kubernetes/kubernetes/pull/104743#issuecomment-916175110 summarizes why that fix is necessary - we don't want people to use static UIDs with static pods, and we want the "obvious correct behavior".  All static pod updates look like force deletion on the apiserver to the kubelet - the old version disappears, the new version is held on the kubelet until the old version completes.  Therefore, reverting #104743 would allow static pods without static uids to be reentrant, which is surprising.

The attached fix looks correct, and after reviewing the test it also seems to reproduce the problem described.  I had some additional feedback on the issue, but I don't think a revert is appropriate until we determine the fix has further issues or there is another issue buried underneath it.

**Comment by rphillips** [comment]
@smarterclayton Prior to https://github.com/kubernetes/kubernetes/pull/104743 there is logic to prevent static pods of the same [name to be started](https://github.com/kubernetes/kubernetes/blame/1e4cacb57743514ffbc45dcdb1abe5797b4d1a97/pkg/kubelet/pod_workers.go#L381). The name is derived from the [namespace]_[pod name] tuple. This check would prevent the static uid from being re-entrant.

**Comment by gjkim42** [comment]
> @smarterclayton Prior to #104743 there is logic to prevent static pods of the same [name to be started](https://github.com/kubernetes/kubernetes/blame/1e4cacb57743514ffbc45dcdb1abe5797b4d1a97/pkg/kubelet/pod_workers.go#L381). The name is derived from the [namespace]_[pod name] tuple. This check would prevent the static uid from being re-entrant since the block is just derived from the tuple and not the UID.

I guess that the logic has not been enough to prevent the new static pod to be reentrant, and #104743 was intended to guarantee the graceful termination of a static pod.

I can understand that this bug is more urgent to be fixed. However, the graceful termination of a static pod that #104743 wanted to guarantee should be fixed someday unless we decide not to support it.

**Comment by rphillips** [comment]
@gjkim42 I'm confused as to the ordering of events. https://github.com/kubernetes/kubernetes/issues/97722 was opened prior to the pod lifecycle refactor, and I think we handled this case. I believe https://github.com/kubernetes/kubernetes/issues/97722 should have been closed. Is there a known test case that failed?

File: test/e2e_node/mirror_pod_grace_period_test.go [hunk 1]
```diff
@@ -56,61 +53,36 @@ var _ = SIGDescribe("MirrorPodWithGracePeriod", func() {
...
 			err = os.Remove(file)
 			framework.ExpectNoError(err)
 
-			ginkgo.By("wait for the mirror pod to be running for grace period")
-			gomega.Consistently(func() error {
-				return checkMirrorPodRunningWithUID(f.ClientSet, mirrorPodName, ns, uid)
-			}, 19*time.Second, 200*time.Millisecond).Should(gomega.BeNil())
-		})
-
-		ginkgo.It("mirror pod termination should satisfy grace period when static pod is updated [NodeConformance]", func() {
```
Comment (gjkim42) [inline test/e2e_node/mirror_pod_grace_period_test.go:75]: I guess the e2e test of this PR demonstrates the failed case. 
or you can reproduce it using https://github.com/kubernetes/kubernetes/issues/97722#issue-778928060 original issue's description.

**Comment by gjkim42** [comment]
@rphillips 
#97722 was fixed by #98103, and was reproduced again after the pod lifecycle refactor.
#104743 fixed it again later.

**Comment by rphillips** [comment]
@gjkim42 Didn't this patch fix it? https://github.com/kubernetes/kubernetes/commit/d5719800bf044b67a538c679dbd808def36bcf0f

**Comment by gjkim42** [comment]
@rphillips 
No to my knowledge. They are different issues.

One handles the recreation of static pods having the same uid with the same fullname(technically one pod).
The other(#97722) handles the recreation(terminate one and create the other) of static pods having the different uid with the same fullname(technically two pods).

**Comment by rphillips** [comment]
After reviewing the original PR, I noticed this [change in logic](https://github.com/kubernetes/kubernetes/pull/104743/files#diff-c6d21f5481c391a3f32bcfddf45c9c20132b2876e8be9690289474b76b71d809L902-L904). completeTerminating does not cleanup startedStaticPodsByFullname, when it did cleanup terminatingStaticPodFullnames before. I tested a build with [similar cleanup logic](https://github.com/openshift/kubernetes/pull/1157/commits/dbfcc1aefdbf03a366eadf7175444cbbc986fa6a) and it does seem to help the issue.

**Comment by gjkim42** [comment]
> @gjkim42 After reviewing the original PR, I noticed this [change in logic](https://github.com/kubernetes/kubernetes/pull/104743/files#diff-c6d21f5481c391a3f32bcfddf45c9c20132b2876e8be9690289474b76b71d809L902-L904). completeTerminating does not cleanup startedStaticPodsByFullname, when it did cleanup terminatingStaticPodFullnames before. I tested a build with [similar cleanup logic](https://github.com/openshift/kubernetes/pull/1157/commits/dbfcc1aefdbf03a366eadf7175444cbbc986fa6a) and it does seem to help the issue.

@rphillips 
I think the logic is technically the same as it was. [This logic](https://github.com/kubernetes/kubernetes/pull/104743/files#diff-c6d21f5481c391a3f32bcfddf45c9c20132b2876e8be9690289474b76b71d809L501-L516) complements the change.
Removing a static pod from startedStaticPodsByFullname in completeTerminating function is not what we want to do.

**Comment by rphillips** [comment]
closing in favor of https://github.com/kubernetes/kubernetes/pull/107900

**Pull request closed without merging**